Add basic message filtering in the SI

zenmonkeykstop opened the pull request:
## Status

 Work in progress (tests required)

## Description of Changes

Fixes #6297.
Fixes #6298.

adds options in the JI instance config, and in the SI,  to:
- set a minimum message length for source's initial messages, as per #6297 
- reject initial messages that consist of the source's codename, as per #6298

## Testing
- check out this branch and run `make dev`

### minimum message length
- in the JI, navigate to the Instance Config page via the Admin page
  - [x] verify that the Submission Preferences  section includes a "Prevent sources from sending initial messages shorter than the minimum required length" option, currently unchecked and with the length field set to 0.
- check the "Prevent sources from sending initial messages shorter than..." checkbox but do not set a length. Click **Update Submission Preferences**
  - [x] verify that the checkbox is unchecked and an error message is displayed asking you to set a length.
 - check the "Prevent sources from sending initial messages shorter than..." checkbox and set a length of between 5 and 20 chars. Click **Update Submission Preferences**
   - [x] verify that the checkbox is checked, the length you chose is set, and a success message is displayed.
- Create a new source account on the SI, navigating through to the /lookup page
  - [x] verify that a notice is displayed under the message textfield like "If you are only sending a message, it must be at least N characters long" where N is the length you chose above
  - [x] Verify that if you try to submit  too short a message, the submission fails with a flashed error like "Your first message must be at least N characters long"
  - [x] Verify that a message N characters long can be submitted successfully
  - [x] after a successful first submission, verify that the  notice under the text field is no longer displayed and you can submit message shorter than N chars
  - [x] in the JI, uncheck the "Prevent sources from sending initial messages shorter than..." checkbox,, click **Update Submission Prefs** and verify that the length is set to zero and a success message is displayed
  - [x] Create a new source account in the SI and verify that no message length limit is mentioned or set

### codename messages

- in the JI, navigate to the Instance Config page via the Admin page
  - [x] verify that the Submission Preferences  section includes a "Prevent sources from submitting their codename as an initial message." option, currently unchecked.
- check the "Prevent sources from submitting their codename as an initial message" checkbox. Click **Update Submission Preferences**
   - [x] verify that the checkbox is checked and a success message is displayed.
 - Create a new source account on the SI, navigating through to the /lookup page
   - [x] Enter the codename as a message and click **Submit** - verify that the message is not submitted and an error is flashed like "Please do not submit your codename.."
   - [x] Enter anything else as a message and click **Submit** - verify that the message was submitted correctly
   - [x]  Enter your codename again and click **Submit** - verify that the message was submitted correctly.

## Deployment

DB migration script required, but new functionality is disabled by default.

## Checklist

### If you made changes to the server application code:

- [ ] Linting (`make lint`) and tests (`make test`) pass in the development container

### If you made non-trivial code changes:

- [ ] I have written a test plan and validated it for this PR

Choose one of the following:

- [ ] I have opened a PR in the [docs repo](https://github.com/freedomofpress/securedrop-docs) for these changes, or will do so later
- [x] I would appreciate help with the documentation
- [ ] These changes do not require documentation

zenmonkeykstop committed:
Added basic message checks for length and codename in Source Interface.

- Updated instance_config table, removing partial index on valid_until field
  that was used to force NULL values to be unique, and using unix epoch start
  as the valid_until value for the active configuration instead. (Partial indexes
  are not fully supported by the Alembic version currently in use.)

- Added optional message filtering:
  - A minimum length can now be set for a source's initial message. Messages shorter
    than this length are rejected with an error in the SI. This feature is managed
    via the Admin interface and is disabled by default.

  - Initial messages can now be rejected if their content matches the source's
    codename. This feature is managed via the Admin interface and is disabled by
    default.

Fixes #6297.
Fixes #6298.

legoktm commented:
Do we also want to use `<textarea minlength="...">`? https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea#attr-minlength

zenmonkeykstop commented:
> Do we also want to use `<textarea minlength="...">`? https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea#attr-minlength

Yep, that'd be a nice addition.

zenmonkeykstop commented:
Looking at that minlength addition with @eaon, it occurred to me that it would prevent initial submissions with a file attachment but no message (which are currently allowed). So, in the absence of client-side logic to allow for that case I think we can't include it except for the edge case upon an edge case of message-only configs. Length checks will be server-side only.

eloquence commented:
Thanks for all the hard work on this! A couple of initial UX notes:

## Organization of new preferences is potentially confusing

![Screenshot from 2022-03-03 17-46-00](https://user-images.githubusercontent.com/213636/156683565-8b4b13c5-1345-4a9b-a470-e7cda24c3fdd.png)

There are IMO a few UX issues here:
- Having the user have to set a "0" value to enable or disable a preference makes it harder for them to follow a single process for all yes/no preferences, to quickly parse the state of preferences, and to know what the previous state was when the feature is disabled.
- One way to avoid this would be to have a "Prevent initial messages below a minimum character length" preference, with the actual character length either as an inline field, or as a separate, indented sub-option.
- Because other preferences follow the "Prevent .." format, the fact that the "Minimum message length" preference does not, can make it feel like it is associated with the previous preference.
- The actual input control only needs to accommodate 4 digits, so it can be far less wide.
## Error message scrolled out of view

Can you reproduce this? If I configure a minimum length and submit a message that gets blocked, the error message is not scrolled within the viewport:
![Screenshot from 2022-03-03 17-45-45](https://user-images.githubusercontent.com/213636/156683453-05acaf45-d859-47fb-84c0-a4f059bf0b92.png)

I don't see this behavior with error messages in `develop` (e.g., when attempting to send a blank message).

zenmonkeykstop commented:
All seems doable, had planned to restrict that length field size via CSS, could probably just add an attribute directly as well.

I'm thinking that the extra para in the SI about message lengths might be what's pushing the flash error out of viewport - for neatness sake I was already thinking of moving that into the placeholder text in the textarea instead. Will push up something tomorrow.

zenmonkeykstop commented:
@eloquence a revision with some UX-related changes addressing your points above is working its way thru CI. Note that it does *not*  update the placeholder text in the submission textarea as I mentioned above - instead the same design language/placement as is currently used for file size limits was applied.

eloquence commented:
Thanks for the changes, it's coming together nicely. I'm still noticing the issue with the error message not being scrolled into view, not sure if you've been able to reproduce / had a chance to debug that one.

The new preferences form seems to work well. I think it would be nice if it remembered the previous value of the form field even after you disable the feature, but that's definitely a nice-to-have.

Regarding the UX changes, I would suggest tweaking as follows:
![Screenshot from 2022-03-07 17-14-16](https://user-images.githubusercontent.com/213636/157146313-39f21e79-9de6-49ff-a69d-a1d49ff436bd.png)

Tweaks:
- No background color (the gray color can suggest a "disabled" look, and does not IMO meaningfully add to the grouping; with the new wording, it seems clearer to me that these are distinct preferences)
- Narrower input
- Changed "minimum required" to just "minimum" ("prevent" in combination with "minimum" are IMO sufficiently clear)
- Avoid repetition in the content hint, style it a bit differently from the field label

Regarding the minimum length hint, I would suggest also slightly tweaking it:

> If your first submission is message-only, it must be at least 5 characters long.

->

> If you are only sending a message, it must be at least 5 characters long.

- Avoid the awkward construction "message-only"
- Keep it simple, trying to communicate too many qualifiers means the message is more likely to be ignored as "fine print".

zenmonkeykstop commented:
@eloquence further tweaks applied.

eaon commented on securedrop/sass/source.sass:
```diff
@@ -255,6 +255,11 @@ p#max-file-size
+p#min-msg-length
+  font-size: 10px
+  font-style: italic
+  margin-bottom: 0
```
I think this margin could be set to 12px - `#max-file-size` after which this is modeled does not have a bottom margin, but it's wrapped in an element with a 12px padding all around, which isn't really suitable for the `textarea` side of things

zenmonkeykstop replied:
as you have spoken, so has it been actioned.

eloquence commented:
Thanks for the additional tweaks! Two remaining issues I'm noticing:

-  The error message if you don't set the integer makes direct reference to internal field names, presumably because it's standard built-in validation logic. (`Preferences not updated. The "min_message_length" field is required when "prevent_short_messages" is set.`) Can we customize that, e.g., "To configure a minimum message length, you must set the  required number of characters." ?
- I'm still seeing the issue with the flash message scrolling out of view.

Otherwise it LGTM, and potentially mergeable for 2.3.0 (I'm still queasy about the min-length change, but I'm OK with it as an opt-in to learn more; as discussed, we may want to at least tweak language on `/generate` to accommodate the codename blocking feature).

zenmonkeykstop commented:
re error, It's actually a custom validator added while fixing the OTP checks - lemme just throw an optional custom message in there real quick...

I'm not seeing the same issue with the flash message and the viewport. Will try to repro.

legoktm commented on securedrop/alembic/versions/55f26cd22043_remove_partial_index_on_valid_until_set_.py:
```diff
@@ -0,0 +1,57 @@
+    # valid_until will be non-nullable after batch, so set existing nulls to
+    # the new default unix epoch datetime
+    op.execute(sa.text(
+        f'UPDATE OR IGNORE instance_config SET '  # nosec
```
We don't need fstrings here since we're not actually interpolating anything, which probably fixes the security warning too.

zenmonkeykstop replied:
fixed

legoktm commented on securedrop/alembic/versions/55f26cd22043_remove_partial_index_on_valid_until_set_.py:
```diff
@@ -0,0 +1,57 @@
+
+    # valid_until is nullable again, set entries with unix epoch datetime value to NULL
+    op.execute(sa.text(
+        f'UPDATE OR IGNORE instance_config SET '  # nosec
```
Same, no f-strings needed

zenmonkeykstop replied:
fixed.

legoktm commented on securedrop/journalist_app/admin.py:
```diff
@@ -36,9 +36,18 @@ def index() -> str:
     @admin_required
     def manage_config() -> Union[str, werkzeug.Response]:
-        # The UI prompt ("prevent") is the opposite of the setting ("allow"):
+        # The UI document upload prompt ("prevent") is the opposite of the setting ("allow")
```
Should this comment be moved to be next to the `prevent_document_uploads=not...` line?

zenmonkeykstop replied:
duplicate comment?

legoktm commented on securedrop/journalist_app/admin.py:
```diff
@@ -36,9 +36,18 @@ def index() -> str:
     @admin_required
     def manage_config() -> Union[str, werkzeug.Response]:
-        # The UI prompt ("prevent") is the opposite of the setting ("allow"):
+        # The UI document upload prompt ("prevent") is the opposite of the setting ("allow")
```
Should this comment be moved to be next to the `prevent_document_uploads=not...` line?

zenmonkeykstop replied:
sure.

legoktm commented on securedrop/journalist_app/admin.py:
```diff
@@ -36,9 +36,18 @@ def index() -> str:
-        # The UI prompt ("prevent") is the opposite of the setting ("allow"):
+        # The UI document upload prompt ("prevent") is the opposite of the setting ("allow")
+        if InstanceConfig.get_default().initial_message_min_len > 0:
+            prevent_short_messages = True
```
This could be one line, `prevent_short_messages = InstanceConfig.get_default().initial_message_min_len > 0`

zenmonkeykstop replied:
Prefer to leave it as is for readability.

legoktm commented on securedrop/journalist_app/admin.py:
```diff
@@ -75,9 +85,27 @@ def update_submission_preferences() -> Optional[werkzeug.Response]:
+
+                    if isinstance(msg_length, int):
+                        int_length = msg_length
+                    elif isinstance(msg_length, str):
```
How do you hit this case? I wasn't able to figure out how to reach it, even when hacking the HTML to post something like "aaa".

When I did submit non-int data, wtforms filled it with None, so then the RequiredIf validator tripped and displayed an error message.

zenmonkeykstop replied:
This is just paranoia on my part, it probably is unnecessary.

zenmonkeykstop replied:
(nuked)

legoktm commented on securedrop/alembic/versions/55f26cd22043_remove_partial_index_on_valid_until_set_.py:
```diff
@@ -0,0 +1,57 @@
+        f'valid_until=:epoch WHERE valid_until IS NULL;'
+    ).bindparams(epoch=datetime.fromtimestamp(0)))
+    with op.batch_alter_table('instance_config', schema=None) as batch_op:
+        batch_op.alter_column('valid_until',
```
Don't we need something to make this unique?

zenmonkeykstop replied:
It's already unique - the gotcha is that uniqueness is not enforced on NULL values in SQL.

legoktm replied:
Ack!

legoktm commented on securedrop/tests/migrations/migration_55f26cd22043.py:
```diff
@@ -0,0 +1,91 @@
+        assert index_definition in get_schema(self.app)
+
+        with self.app.app_context():
+            with pytest.raises(sqlalchemy.exc.OperationalError):
```
The `raises()` should be inside the for loop, otherwise after it raises on the first one it won't run the second query.

zenmonkeykstop replied:
aye, moved.

legoktm commented on securedrop/tests/migrations/migration_55f26cd22043.py:
```diff
@@ -0,0 +1,91 @@
+                     ]:
+
+                    result = db.engine.execute(sqlalchemy.text(query)).fetchall()
+                    assert len(result) == 0
```
This seems unnecessary? Since it'll raise.

zenmonkeykstop replied:
aye, deleted.

eloquence commented:
The viewport issue seems very dependent on the window size. At a high vertical resolution, this is the kind of worst case that I'm seeing:
![flash](https://user-images.githubusercontent.com/213636/157574307-f2edd0bb-8234-49dd-8bd0-bb950c30962a.gif)

It's possible that attempting to anchor to the flash message is more trouble than it's worth.

cfm commented:
<https://github.com/freedomofpress/securedrop/pull/6306#issuecomment-1063580054>:

> The viewport issue seems very dependent on the window size. At a high vertical resolution, this is the kind of worst case that I'm seeing: [...]
> 
> It's possible that attempting to anchor to the flash message is more trouble than it's worth.

This originated in #6041 as an [accessibility feature](https://docs.google.com/document/d/1jZlKco2YgOrrI4PmTkG_Q8lKuV5VxCds/edit?disco=AAAAIplLpuY) (see especially top of page 53), so that screen-readers can provide an equivalent to `#flashed`'s visual emphasis.  As much as I wish the browser's behavior here didn't force us to choose which experience (read: class of users) to privilege—at least without the benefit of the recommended JavaScript implementation—I agree that we shouldn't let this desideratum compromise the visual UX.

legoktm commented:
I'll double check with codecov after CI finishes, but I think all the new code should be covered by tests now.

eaon commented:
Regarding `#flashed` viewport jump issue, I've not paid enough attention to this behaviour before, but I just confirmed that this happens to me on the demo instance. I opened #6333 so we can separate that problem from the PR

legoktm commented:
Test plan all checks out (and pointed out a gap in the tests I wrote)!

zenmonkeykstop commented:
(for whoever ends up merging, this can probably be rebased into a single commit first)

eloquence commented on securedrop/source_app/main.py:
```diff
@@ -192,6 +200,23 @@ def submit(logged_in_source: SourceUser) -> werkzeug.Response:
+                flash(gettext(
+                    "Your first message must be at least {} characters long.".format(min_len)),
+                    "error")
+                return redirect(f"{url_for('main.lookup')}#flashed")
```
Per #6336 I think we'll want to now redirect without the #flashed fragment.

eloquence commented on securedrop/source_app/main.py:
```diff
@@ -192,6 +200,23 @@ def submit(logged_in_source: SourceUser) -> werkzeug.Response:
+                    escape(gettext("Keep your codename secret, and use it to log in later"
+                                   " to check for replies."))
+                    )), "error")
+                return redirect(f"{url_for('main.lookup')}#flashed")
```
As above.

eloquence commented:
I've pushed up a branch rebased to `develop`, with the anchor tweaks above, to `add-message-filtering-redux`, in case you want to use it. (I did not want to force push to this branch in case you have local changes.)

legoktm commented:
I'm assuming Kev had nothing left locally, so I'm going to cherry-pick 540a693bcdce595a2a8764c631bba2dfb0c50681 over to this branch, push, then squash+rebase it all and push again.

legoktm reviewed (dismissed):
I think we're set to land this but I'd like someone else to confirm that they reviewed the test cases I wrote (so I'm not merging them without review), and double check my squash and commit message.

cfm reviewed (dismissed):
Reviewed test cases and rebase/squash at @legoktm's request.  All looks good to me except for the following nit about the commit message on the squashed 500531deea0372c332b55f911fda5988b828fb64:

> When trying to do the schema change for the InstanceConfig table, it was
> discovered that SQLAlchemy/Alembic don't support the partial index on
> the table, so it's dropped and the valid_until column is now
> nullable=False. The current/active config is represented by a
> valid_until of the unix epoch (0).

On first reading, if one isn't already familiar with this problem, it might be unclear whether the clause beginning "so it's dropped" describes (a) SQLAlchemy's behavior in not supporting the partial index *prior to* this commit or (b) the fix made *by* this commit.  Up to you, @legoktm, whether it's worth revising, but since the commit (once merged) is the immutable thing here I thought I'd mention it.  :-)

zenmonkeykstop commented:
> On first reading, if one isn't already familiar with this problem, it might be unclear whether the clause beginning "so it's dropped" describes (a) SQLAlchemy's behavior in not supporting the partial index _prior to_ this commit or (b) the fix made _by_ this commit. Up to you, @legoktm, whether it's worth revising, but since the commit (once merged) is the immutable thing here I thought I'd mention it. :-)

Agreed, reworded.

conorsch reviewed (approved):
Re-blessing for merge after requested changes were made. :sunglasses:
